src/apps/applib/lib/utility.c

```diff
@@ -21,13 +21,159 @@
 #include <assert.h>
 #include <inttypes.h>
 #include <stackAlloc.h>
+#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include "coordijk.h"
 #include "geoCoord.h"
 #include "h3Index.h"
 #include "h3api.h"
 
+/**
+ * Parse command line arguments.
+ *
+ * Uses the provided arguments to populate argument values.
+ *
+ * Returns non-zero if all required arguments are not present, an argument fails
+ * to parse, is missing its associated value, or arguments are specified more
+ * than once.
+ *
+ * @param argc argc from main
+ * @param argv argv from main
+ * @param numArgs Number of elements in the args array
+ * @param args Each argument to parse.
+ * @param errorMessage Error message to display, if returning non-zero. Caller
+ * must free this.
+ * @param errorDetail Additional error details, if returning non-zero. May be
+ * null, caller must free this.
+ * @return 0 if argument parsing succeeded, otherwise non-0.
+ */
+int parseArgs(int argc, char* argv[], int numArgs, const Arg* args,
+              char** errorMessage, char** errorDetail) {
+    bool* foundArgs = calloc(numArgs, sizeof(bool));
+    // Whether help was found and required arguments do not need to be checked
+    bool foundHelp = false;
+
+    for (int i = 1; i < argc; i++) {
+        bool foundMatch = false;
+
+        for (int j = 0; j < numArgs; j++) {
+            const char* argName = NULL;
+            bool isMatch = false;
+            for (int k = 0; k < NUM_ARG_NAMES; k++) {
+                if (args[j].names[k] == NULL) continue;
+
+                if (strcmp(argv[i], args[j].names[k]) == 0) {
+                    argName = args[j].names[k];
+                    isMatch = true;
+                    break;
+                }
+            }
+            if (!isMatch) continue;
+
+            if (foundArgs[j]) {
+                free(foundArgs);
+                *errorMessage = strdup("Argument specified multiple times");
+                *errorDetail = strdup(argName);
+                return 1;
+            }
+
+            if (args[j].valuePresent != NULL) {
+                *args[j].valuePresent = true;
+            }
+            if (args[j].scanFormat != NULL) {
+                i++;
+                if (i >= argc) {
+                    free(foundArgs);
+                    *errorMessage = strdup("Argument value not present");
+                    *errorDetail = strdup(argName);
+                    return 2;
+                }
+
+                if (!sscanf(argv[i], args[j].scanFormat, args[j].value)) {
+                    free(foundArgs);
+                    *errorMessage = strdup("Failed to parse argument");
+                    *errorDetail = strdup(argName);
+                    return 3;
+                }
+            }
```

Comment:
I'm not sure we could reliably differentiate that from another argument being present next. It will hit an error condition when the next argument is parsed.